modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/Resource.kt

```diff
@@ -0,0 +1,217 @@
+package arrow.effects
+
+import arrow.Kind
+import arrow.core.andThen
+import arrow.effects.typeclasses.Bracket
+import arrow.effects.typeclasses.ExitCase
+import arrow.higherkind
+import arrow.typeclasses.Monoid
+import arrow.typeclasses.Semigroup
+
+/**
+ * ank_macro_hierarchy(arrow.effects.Resource)
+ *
+ * [Resource] models resource allocation and releasing. It is especially useful when multiple resources that depend on each other
+ *  need to be acquired and later released in reverse order.
+ * When a resource is created one can make use of [invoke] to run a computation with the resource. The finalizers are then
+ *  guaranteed to run afterwards in reverse order of acquisition.
+ *
+ * Consider the following use case:
+ * ```kotlin:ank:playground
+ * import arrow.effects.IO
+ * import arrow.effects.extensions.io.fx.fx
+ *
+ * object Consumer
+ * object Handle
+ *
+ * class Service(val handle: Handle, val consumer: Consumer)
+ *
+ * fun createConsumer(): IO<Consumer> = IO { println("Creating consumer"); Consumer }
+ * fun createDBHandle(): IO<Handle> = IO { println("Creating db handle"); Handle }
+ * fun createFancyService(consumer: Consumer, handle: Handle): IO<Service> = IO { println("Creating service"); Service(handle, consumer) }
+ *
+ * fun closeConsumer(consumer: Consumer): IO<Unit> = IO { println("Closed consumer") }
+ * fun closeDBHandle(handle: Handle): IO<Unit> = IO { println("Closed db handle") }
+ * fun shutDownFanceService(service: Service): IO<Unit> = IO { println("Closed service") }
+ *
+ * //sampleStart
+ * val program = fx {
+ *   val consumer = createConsumer().bind()
```

Comment:
Can we use ! instead of bind() that is the preferred way in fx

Reply:
Changed. That only goes for `fx` blocks right? I left the resource binding as is for now

Reply:
Oh and something I forgot: Don't ank:playground snippets require all code despite actually compiling with just the important bits? If so I need to add some stuff before merge :/